flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java

```diff
@@ -0,0 +1,429 @@
+ * that reports it is available.
+ *
+ * <p>The sorter uses binary comparison of keys, which are extracted and serialized when received
+ * from the chained input. Moreover the timestamps of incoming records are used for secondary ordering.
+ * For the comparison it uses either {@link FixedLengthByteKeyComparator} if the length of the
+ * serialized key is constant, or {@link VariableLengthByteKeyComparator} otherwise.
+ *
+ * <p>Watermarks, stream statuses, nor latency markers are propagated downstream as they do not make
+ * sense with buffered records. The input emits a MAX_WATERMARK after all records.
+ */
+public final class MultiInputSortingDataInput<IN, K> implements StreamTaskInput<IN> {
+	private final int idx;
+	private final StreamTaskInput<IN> wrappedInput;
+	private final PushSorter<Tuple2<byte[], StreamRecord<IN>>> sorter;
+	private final CommonContext commonContext;
+	private final SortingPhaseDataOutput sortingPhaseDataOutput = new SortingPhaseDataOutput();
+
+	private final KeySelector<IN, K> keySelector;
+	private final TypeSerializer<K> keySerializer;
+	private final DataOutputSerializer dataOutputSerializer;
+
+	private MutableObjectIterator<Tuple2<byte[], StreamRecord<IN>>> sortedInput;
+	private long seenWatermark = Long.MIN_VALUE;
+
+	private MultiInputSortingDataInput(
+			CommonContext commonContext,
+			StreamTaskInput<IN> wrappedInput,
+			int inputIdx,
+			PushSorter<Tuple2<byte[], StreamRecord<IN>>> sorter,
+			KeySelector<IN, K> keySelector,
+			TypeSerializer<K> keySerializer,
+			DataOutputSerializer dataOutputSerializer) {
+		this.wrappedInput = wrappedInput;
+		this.idx = inputIdx;
+		this.commonContext = commonContext;
+		this.sorter = sorter;
+		this.keySelector = keySelector;
+		this.keySerializer = keySerializer;
+		this.dataOutputSerializer = dataOutputSerializer;
+	}
+
+	public static <K> StreamTaskInput<?>[] wrapInputs(
+			AbstractInvokable containingTask,
+			StreamTaskInput<Object>[] inputs,
+			KeySelector<Object, K>[] keySelectors,
+			TypeSerializer<Object>[] inputSerializers,
+			TypeSerializer<K> keySerializer,
+			MemoryManager memoryManager,
+			IOManager ioManager,
+			boolean objectReuse,
+			double managedMemoryFraction,
+			Configuration jobConfiguration) {
+		int keyLength = keySerializer.getLength();
+		final TypeComparator<Tuple2<byte[], StreamRecord<Object>>> comparator;
+		DataOutputSerializer dataOutputSerializer;
+		if (keyLength > 0) {
+			dataOutputSerializer = new DataOutputSerializer(keyLength);
+			comparator = new FixedLengthByteKeyComparator<>(keyLength);
+		} else {
+			dataOutputSerializer = new DataOutputSerializer(64);
+			comparator = new VariableLengthByteKeyComparator<>();
+		}
+
+		int numberOfInputs = inputs.length;
+		CommonContext commonContext = new CommonContext(numberOfInputs);
+		return IntStream.range(0, numberOfInputs)
+			.mapToObj(
+				idx -> {
+					try {
+					KeyAndValueSerializer<Object> keyAndValueSerializer = new KeyAndValueSerializer<>(
```

Comment:
Indentation seems off here. (Another point for having an automatic formatting tool... 😅)